include/mbgl/util/run_loop.hpp

```diff
@@ -74,22 +84,34 @@ class RunLoop : public Scheduler,
 private:
     MBGL_STORE_THREAD(tid)
 
-    using Queue = std::queue<std::shared_ptr<WorkTask>>;
+    using Queue = std::list<std::shared_ptr<WorkTask>>;
 
-    void push(std::shared_ptr<WorkTask>);
+    // Wakes up the RunLoop so that it starts processing items in the queue.
+    void wake();
 
-    void withMutex(std::function<void()>&& fn) {
+    // Adds a WorkTask to the queue, and wakes it up.
+    void push(Priority priority, std::shared_ptr<WorkTask> task) {
         std::lock_guard<std::mutex> lock(mutex);
-        fn();
+        if (priority == Priority::High) {
+            // We're always placing high priority tasks at the front. This means it's not really a
```

Comment:
Addressed this by having two separate queues and checking the high priority one first. I chose this implementation over a `priority_queue` since it removes the need to sort items upon insertion.